libvast/builtins/operators/tail.cpp

```diff
@@ -0,0 +1,122 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/concept/parseable/numeric/integral.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/error.hpp>
+#include <vast/legacy_pipeline.hpp>
+#include <vast/logger.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/table_slice.hpp>
+
+#include <arrow/type.h>
+
+namespace vast::plugins::tail {
+
+namespace {
+
+class tail_operator final : public crtp_operator<tail_operator> {
+public:
+  explicit tail_operator(uint64_t limit) : limit_{limit} {
+  }
+
+  auto operator()(generator<table_slice> input) const
+    -> generator<table_slice> {
+    auto buffer = std::vector<table_slice>{};
+    auto total_buffered = size_t{0};
+    for (auto&& slice : input) {
+      total_buffered += slice.rows();
+      buffer.push_back(std::move(slice));
+      if (total_buffered - buffer.front().rows() >= limit_) {
+        total_buffered -= buffer.front().rows();
+        buffer.erase(buffer.begin());
+      }
+      co_yield {};
+    }
+    if (buffer.empty())
+      co_return;
+    buffer.front() = vast::tail(buffer.front(), buffer.front().rows()
+                                                  - (total_buffered - limit_));
+    for (auto&& slice : std::move(buffer))
+      co_yield std::move(slice);
+  }
+
+  auto to_string() const -> std::string override {
+    return fmt::format("tail {}", limit_);
+  }
+
+private:
+  uint64_t limit_;
+};
+
+class plugin final : public virtual pipeline_operator_plugin,
+                     public virtual operator_plugin {
+public:
+  // plugin API
+  auto initialize([[maybe_unused]] const record& plugin_config,
+                  [[maybe_unused]] const record& global_config)
+    -> caf::error override {
+    return {};
+  }
+
+  [[nodiscard]] auto name() const -> std::string override {
+    return "tail";
+  };
+
+  // transform plugin API
+  [[nodiscard]] auto make_pipeline_operator(const record&) const
+    -> caf::expected<std::unique_ptr<legacy_pipeline_operator>> override {
+    return caf::make_error(ec::unimplemented,
+                           "the tail operator is not available in the YAML "
+                           "operator syntax");
+  }
+
+  [[nodiscard]] auto make_pipeline_operator(std::string_view pipeline) const
+    -> std::pair<
+      std::string_view,
+      caf::expected<std::unique_ptr<legacy_pipeline_operator>>> override {
+    const auto* f = pipeline.begin();
+    const auto* const l = pipeline.end();
+    return {
+      std::string_view{f, l},
+      caf::make_error(ec::unimplemented,
+                      "the tail operator is not available in the old version "
+                      "of the pipeline API"),
+    };
+  }
```

Comment:
I don't think you need to make this one of the old operator plugins if you're not implementing it anyways.

```suggestion
class plugin final : public virtual operator_plugin {
public:
  // plugin API
  auto initialize([[maybe_unused]] const record& plugin_config,
                  [[maybe_unused]] const record& global_config)
    -> caf::error override {
    return {};
  }

  [[nodiscard]] auto name() const -> std::string override {
    return "tail";
  };
```